util/network-devp2p/src/discovery.rs

```diff
@@ -119,87 +149,81 @@ impl Discovery {
 			discovery_id: NodeId::new(),
 			discovery_nodes: HashSet::new(),
 			node_buckets: (0..ADDRESS_BITS).map(|_| NodeBucket::new()).collect(),
+			in_flight_requests: HashMap::new(),
+			expiring_pings: VecDeque::new(),
+			expiring_finds: VecDeque::new(),
 			udp_socket: socket,
 			send_queue: VecDeque::new(),
 			check_timestamps: true,
 			adding_nodes: Vec::new(),
 			ip_filter: ip_filter,
+			request_backoff: REQUEST_BACKOFF.iter().map(|s| Duration::from_secs(*s)).collect(),
 		}
 	}
 
 	/// Add a new node to discovery table. Pings the node.
 	pub fn add_node(&mut self, e: NodeEntry) {
-		if self.is_allowed(&e) {
-			let endpoint = e.endpoint.clone();
-			self.update_node(e);
-			self.ping(&endpoint);
+		// If distance returns None, then we are trying to add ourself.
+		let id_hash = keccak(e.id);
+		if let Some(dist) = Discovery::distance(&self.id_hash, &id_hash) {
+			if self.node_buckets[dist].nodes.iter().any(|n| n.id_hash == id_hash) {
+				return;
+			}
+			self.try_ping(e);
 		}
 	}
 
 	/// Add a list of nodes. Pings a few nodes each round
-	pub fn add_node_list(&mut self, nodes: Vec<NodeEntry>) {
-		self.adding_nodes = nodes;
-		self.update_new_nodes();
+	pub fn add_node_list(&mut self, mut nodes: Vec<NodeEntry>) {
+		for node in nodes {
+			self.add_node(node);
+		}
 	}
 
 	/// Add a list of known nodes to the table.
 	pub fn init_node_list(&mut self, mut nodes: Vec<NodeEntry>) {
-		for n in nodes.drain(..) {
+		for n in nodes {
```

Comment:
The `nodes` parameter here and in `add_node_list` does not need to be declared as `mut`.